File: packages/sdk/src/resources/loyalty.ts
import type { TheatricalHTTPClient } from '../http/client';
import type {
  LoyaltyMember,
  PointsTransaction,
  RedemptionOption,
  RedeemPointsInput,
  PointsHistoryFilter,
} from '../types/loyalty';
import type { PaginatedResponse } from '../types/pagination';

/**
 * Loyalty resource — member management, points, tiers.
 *
 * Wraps the Vista OCAPI loyalty endpoints for full member lifecycle management:
 * authentication, balance queries, transaction history, and points redemption.
 */
export class LoyaltyResource {
  constructor(private readonly http: TheatricalHTTPClient) {}

  /**
   * Fetch a loyalty member by their unique ID.
   *
   * @param memberId - Vista loyalty member identifier
   */
  async getMember(memberId: string): Promise<LoyaltyMember> {
    return this.http.get<LoyaltyMember>(`/ocapi/v1/loyalty/members/${memberId}`);
  }

  /**
   * Authenticate a member using their email address and password (cookie-based
   * session). Returns the authenticated member record on success.
   *
   * @param email    - Member's registered email
   * @param password - Account password (transmitted over TLS)
   */
  async authenticate(email: string, password: string): Promise<LoyaltyMember> {
    return this.http.post<LoyaltyMember>('/ocapi/v1/loyalty/authenticate', {
      body: { email, password },
    });
  }

  /**
   * Return the current redeemable points balance for a member.
   *
   * @param memberId - Vista loyalty member identifier
   */
  async getPointsBalance(memberId: string): Promise<{ points: number; lifetimePoints: number }> {
    return this.http.get<{ points: number; lifetimePoints: number }>(
      `/ocapi/v1/loyalty/members/${memberId}/points`,
    );
  }

  /**
   * Retrieve a paginated transaction history for a member.
   *
   * @param memberId - Vista loyalty member identifier
   * @param filter   - Optional date range, type, and pagination filters
   */
  async getHistory(
    memberId: string,
    filter?: PointsHistoryFilter,
  ): Promise<PaginatedResponse<PointsTransaction>> {
    return this.http.get<PaginatedResponse<PointsTransaction>>(
      `/ocapi/v1/loyalty/members/${memberId}/history`,
      { params: filter as Record<string, string | number | boolean | undefined> | undefined },
    );
  }

  /**
   * List available redemption options for the loyalty program.
   * Options are filtered to those the member is eligible for based on their
   * tier and current points balance.
   *
   * @param memberId - Vista loyalty member identifier
   */
  async listRedemptionOptions(memberId: string): Promise<RedemptionOption[]> {
    return this.http.get<RedemptionOption[]>(
      `/ocapi/v1/loyalty/members/${memberId}/redemptions`,
    );
  }

  /**
   * Redeem points against a catalog option for a member.
   * Deducts points from the member's balance and applies the benefit.
   *
   * @param memberId - Vista loyalty member identifier
   * @param input    - Redemption option, optional order linkage, and quantity
   */
  async redeemPoints(memberId: string, input: RedeemPointsInput): Promise<PointsTransaction> {
    return this.http.post<PointsTransaction>(
      `/ocapi/v1/loyalty/members/${memberId}/redeem`,
      { body: input },
    );
  }
}

File: packages/sdk-py/theatrical/resources/loyalty.py
"""Loyalty resource — member management, points, tiers."""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

from pydantic import TypeAdapter

from theatrical.types.loyalty import (
    LoyaltyMember,
    PointsBalance,
    PointsHistoryFilter,
    PointsTransaction,
    RedeemPointsInput,
    RedemptionOption,
)
from theatrical.types.pagination import PaginatedResponse

if TYPE_CHECKING:
    from theatrical.http.client import TheatricalHttpProtocol

_member_adapter = TypeAdapter(LoyaltyMember)
_points_balance_adapter = TypeAdapter(PointsBalance)
_paginated_tx_adapter = TypeAdapter(PaginatedResponse[PointsTransaction])
_redemption_list_adapter = TypeAdapter(list[RedemptionOption])
_tx_adapter = TypeAdapter(PointsTransaction)


class LoyaltyResource:
    def __init__(self, http: TheatricalHttpProtocol) -> None:
        self._http = http

    async def get_member(self, member_id: str) -> LoyaltyMember:
        raw = await self._http.get(f"/ocapi/v1/loyalty/members/{member_id}")
        return _member_adapter.validate_python(raw)

    async def authenticate(self, email: str, password: str) -> LoyaltyMember:
        raw = await self._http.post(
            "/ocapi/v1/loyalty/authenticate",
            body={"email": email, "password": password},
        )
        return _member_adapter.validate_python(raw)

    async def get_points_balance(self, member_id: str) -> PointsBalance:
        raw = await self._http.get(f"/ocapi/v1/loyalty/members/{member_id}/points")
        return _points_balance_adapter.validate_python(raw)

    async def get_history(
        self, member_id: str, filters: Optional[PointsHistoryFilter] = None
    ) -> PaginatedResponse[PointsTransaction]:
        params: dict[str, str] | None = None
        if filters:
            params = {}
            d = filters.model_dump(exclude_none=True)
            for key, value in d.items():
                params[key] = str(value) if not isinstance(value, str) else value
        raw = await self._http.get(
            f"/ocapi/v1/loyalty/members/{member_id}/history", params=params
        )
        return _paginated_tx_adapter.validate_python(raw)

    async def list_redemption_options(self, member_id: str) -> list[RedemptionOption]:
        raw = await self._http.get(
            f"/ocapi/v1/loyalty/members/{member_id}/redemptions"
        )
        return _redemption_list_adapter.validate_python(raw)

    async def redeem_points(
        self, member_id: str, input: RedeemPointsInput
    ) -> PointsTransaction:
        raw = await self._http.post(
            f"/ocapi/v1/loyalty/members/{member_id}/redeem",
            body=input.model_dump(exclude_none=True),
        )
        return _tx_adapter.validate_python(raw)
